xapian-applications/omega/postprocess

```diff
@@ -0,0 +1,247 @@
+            qid, did = row[QUERYID], row[DOCID]
+
+            did_to_count = qid_to_clicks[qid]
+
+            # Check if did is already present in did_to_count
+            if did in did_to_count:
+                # Update did_to_count[did]
+                did_to_count[did] += 1
+            else:
+                did_to_count[did] = 1
+
+            qid_to_clicks[qid] = did_to_count
+
+    with open(search_log, 'r') as search_f, open(final_log, 'w+') as final_f:
+        search_reader = csv.reader(search_f)
+        writer = csv.writer(final_f)
+
+        # Add headers to final log file
+        writer.writerow(["QueryID", "Query", "Hits", "Offset", "Clicks"])
+
+        queries = set()
+
+        for row in search_reader:
+            # Skip rows with empty Query string
+            if row[QUERY] == '':
+                continue
+
+            # Avoid duplicate entries
+            if row[QUERY] in queries:
+                continue
+
+            queries.add(row[QUERY])
+
+            # Convert Hitlist from str to list
+            if not row[HITS]:
+                hits = []
+            else:
+                hits = row[HITS]
+                hits = hits.strip().split(',')
+                row[HITS] = hits
+
+            clicklist = hits[:]
+
+            # Update clicklist with click values stored in map.
+            if row[QUERYID] in qid_to_clicks:
+                did_to_count = qid_to_clicks[row[QUERYID]]
+                for index, did in enumerate(clicklist):
+                    if did in did_to_count:
+                        clicklist[index] = '%s:%i' % (did, did_to_count[did])
+                    else:
+                        clicklist[index] = did + ':0'
+            else:
+                for index, did in enumerate(clicklist):
+                    clicklist[index] = did + ':0'
+
+            # Serialise "Hits" and "Clicks"
+            row[HITS] = ','.join(row[HITS])
+            clicklist = ','.join(clicklist)
+
+            row.append(clicklist)
+            writer.writerow(row)
+
+
+def generate_query_file(final_log, query_file):
+    """Generates query file formatted as per Xapian Letor documentation.
+
+    Input Args:
+        final_log (string): Path to final.log file.
+        query_file (string): Path to query.txt file.
+
+    Example (comma-delimited) entries in final.log:
+
+    QueryID,Query,Hits,Offset,Clicks
+    821f03288846297c2cf43c34766a38f7,book,"45,54",0,"45:0,54:2"
+    098f6bcd4621d373cade4e832627b4f6,test,"35,47",0,"35:1,47:0"
+
+    Example (comma-delimited) entries in query.txt:
+
+    821f03288846297c2cf43c34766a38f7,book
+    098f6bcd4621d373cade4e832627b4f6,test
+    """
+    with open(final_log, 'r') as s, open(query_file, 'w+') as w:
+        reader = csv.DictReader(s)
+        writer = csv.writer(w)
+
+        for row in reader:
+            writer.writerow([row['QueryID'], row['Query']])
+
+
+def test_functions():
+    with open('test_search.log', 'w+') as test_search:
```

Comment:
You shouldn't use hard-coded local file paths, which was partly what I wanted you to change in using mocking/patching. If you want to use temporary files, look at `tempfile` to put them in a run-specific, safe, location.